bazel_repositories_priv

```diff
@@ -0,0 +1,292 @@
+    os.environ["HOME"] = f"/home/{user}"
+    os.environ["USER"] = user
+
+
+def find_rootdir(filename_wildcard: str) -> str:
+    """Return the main directory inside the tarball with sources of packaged
+    project.
+    """
+    root_dir = None
+    filename = glob.glob(filename_wildcard)[0]
+    tar = tarfile.open(filename)
+    for tarinfo in tar:
+        root_dir = tarinfo.name
+        break
+    tar.close()
+    return root_dir
+
+
+def bazel_clean(root_dir: str):
+    """Clean Bazel cache.
+    """
+    os.chdir(root_dir)
+    as_user()
+    subprocess.run(["bazel", "clean", "--expunge"], check=True)
+
+
+def sub(conn: m_connection.Connection, netns_name: str, root_dir: str,
+        bazel_args: str, target: str):
+    """The main sub process function which runs Bazel inside a network
+    namespace without Internet connectivity.
+    """
+    log.debug(f"sub target: {target}")
+    os.chdir(root_dir)
+
+    nsfd = os.open(os.path.join("/var/run/netns", netns_name), os.O_RDONLY)
+    libc.setns(nsfd, CLONE_NEWNET)
+
+    as_user()
+
+    args = ["bazel", "fetch", "--repository_cache=../BAZEL_CACHE"]
+    if bazel_args:
+        args += bazel_args.split()
+    args += [target]
+
+    log.debug(f"args: {args}")
+
+    p = subprocess.Popen(args, stdout=subprocess.PIPE,
+                         stderr=subprocess.STDOUT,
+                         universal_newlines=True)
+    stdout = ""
+    for line in iter(p.stdout.readline, ""):
+        line_s = line.strip()
+        log.debug(f"bazel fetch: {line_s}")
+        stdout += line
+    p.stdout.close()
+    returncode = p.wait()
+
+    urls = re.findall(r"https?://(?:[a-zA-Z]|[0-9]|[-_@.&/+])+", stdout)
+    urls = list(dict.fromkeys(urls))
+
+    conn.send((returncode, urls,))
+    conn.close()
+
+
+def chown(pathname: str):
+    """Change the owner of the given file from root to the unprivileged user.
+    """
+    user_passwd = pwd.getpwnam(user)
+    log.debug(f"chown {pathname}")
+    os.chown(pathname, user_passwd.pw_uid, user_passwd.pw_gid)
+    for root, dirs, files in os.walk(pathname):
+        for d in dirs:
+            filename = os.path.join(root, d)
+            log.debug(f"chown {filename}")
+            os.chown(filename, user_passwd.pw_uid, user_passwd.pw_gid)
+        for f in files:
+            filename = os.path.join(root, f)
+            log.debug(f"chown {filename}")
+            os.chown(filename, user_passwd.pw_uid, user_passwd.pw_gid)
+
+
+def download(url: str, dst: str):
+    log.info(f"Downloading {url}")
+    urllib_request.urlretrieve(url, dst)
+    log.info("Download finished")
+
+
+def process_url(url: str, url_path: pathlib.Path, filename: str) -> str:
+    """Download and store the archive from the given URL.
+    """
+    # If two first chars are not alpha, it means that the tarball name is a
+    # version or a commit hash. In that case, append the project name for
+    # readability.
+    if not (filename[0].isalpha() and filename[1].isalpha()):
```

Comment:
The `isalpha()` method returns true if all chars in the string are letters, so you could do this test with an array slice.  Starting at position 0, with a length of 2:
```suggestion
    if not (filename[0:2].isalpha()):
```

Demo:
```python
>>> s='ab123'
>>> s[0:2].isalpha()
True
>>> s[0:3].isalpha()
False
```

Reply:
Thanks! Although I will remove that as well.

Reply:
Well, maybe useful for next time. 😂